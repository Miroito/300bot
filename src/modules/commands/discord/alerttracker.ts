import discord from 'discord.js'
import { Command } from '@commands/CommandHandler'
import { validateArgumentNumber } from '@commands/validators'
import { alertTrackerDatabase } from '@database/alerttracker'

export default new Command<discord.Message>({
  keyword: 'alerttracker',
  description: 'configure the alert tracker',
  help: 'Usage: `{prefix}alerttracker <on/off>` - turns the alert tracker on or off',
  category: 'Admin',
  callback: async ({ args, reply, env, raw }) => {
    if (args.length === 0) {
      return reply(env.command.getHelp(env.handler))
    }
    validateArgumentNumber(args.length, 1)
    const option = args[0]
    if (option !== 'on' && option !== 'off') {
      return reply('Error: The argument must be either "on" or "off".')
    }
    const channel = raw.channel
    if (!(channel instanceof discord.TextChannel)) {
      return reply('Error: This type of channel is not supported.')
    }

    const dbPath = `${channel.guild.id}` as const
    const currentChannelId = alertTrackerDatabase.get(dbPath)

    if (option === 'on') {
      if (channel.id === currentChannelId) {
        return reply('Alert tracker is already on in this channel.')
      } else {
        await alertTrackerDatabase.set(dbPath, channel.id)
        return reply('Alert tracker is now on in this channel.')
      }
    } else {
      if (currentChannelId === undefined) {
        return reply('Alert tracker is already turned off.')
      } else {
        await alertTrackerDatabase.delete(dbPath)
        return reply('Alert tracker is now turned off.')
      }
    }
  },
})
