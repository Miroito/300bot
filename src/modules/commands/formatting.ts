import { Character, CharacterWithOutfitWithLeader } from '@planetside/types'

const factionEmojis = [
  'Unknown Faction',
  '{emoji:faction_logo_vs|VS}',
  '{emoji:faction_logo_nc|NC}',
  '{emoji:faction_logo_tr|TR}',
  '{emoji:faction_logo_ns|NS}',
]

export const formatChacarcterWithFaction = (
  character: Character | CharacterWithOutfitWithLeader,
): string => {
  return `**${character.name.first}** ${
    factionEmojis[Number(character.factionId)]
  }${
    character.factionId === '4' &&
    'outfitMember' in character &&
    character.outfitMember
      ? ` ${
          factionEmojis[Number(character.outfitMember.outfit.leader.factionId)]
        }`
      : ''
  }`
}
