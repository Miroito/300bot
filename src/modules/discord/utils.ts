import discord from 'discord.js'
import { schedule } from '@app/utils/async'
import { log } from '@app/utils/log'
import { reviveDatabase } from '@database/revive'

export const getTextChannel = (
  client: discord.Client,
  id: string,
): discord.TextChannel | null => {
  const channel = client.channels.resolve(id)
  if (channel instanceof discord.TextChannel) {
    return channel
  }
  return null
}

export const getDMChannel = async (
  client: discord.Client,
  id: string,
): Promise<discord.DMChannel | null> => {
  const channel = await client.channels.fetch(id)
  if (channel instanceof discord.DMChannel) {
    return channel
  }
  return null
}

export const formatWithEmojis = (
  channel: discord.Channel,
  message: string,
): string => {
  // eslint-disable-next-line unicorn/no-unsafe-regex
  const emojiExpressionRegex = /{emoji:([^|]*)(?:\|([^}]*))?}/g

  if (channel instanceof discord.TextChannel) {
    return message.replace(
      emojiExpressionRegex,
      (_match, emojiName, fallbackText) => {
        const emoji = channel.guild.emojis.cache.find(
          ({ name }) => name === emojiName,
        )
        if (emoji) {
          return emoji.toString()
        } else if (typeof fallbackText === 'string') {
          return fallbackText
        } else {
          return ''
        }
      },
    )
  } else {
    return message.replace(
      emojiExpressionRegex,
      (_match, _emojiName, fallbackText) => {
        if (typeof fallbackText === 'string') {
          return fallbackText
        } else {
          return ''
        }
      },
    )
  }
}

export const removeReaction = async (
  reaction: discord.MessageReaction,
  user: discord.User,
): Promise<void> => {
  try {
    await reaction.users.remove(user)
  } catch (error) {
    log.error('Error removing reaction:', error)
  }
}

export const reviveMember = async (
  member: discord.GuildMember,
  deadRole: discord.Role,
): Promise<void> => {
  const guildId = member.guild.id
  const userId = member.id

  await member.roles.remove(deadRole)
  await reviveDatabase.delete(`${guildId}.${userId}`)
}

export const killMember = async (
  member: discord.GuildMember,
  deadRole: discord.Role,
  reviveDelayMs: number,
): Promise<void> => {
  const userId = member.id
  const guildId = member.guild.id

  const reviveTime = Date.now() + reviveDelayMs

  await reviveDatabase.set(`${guildId}.${userId}`, reviveTime)
  await member.roles.add(deadRole)

  void schedule(async () => {
    try {
      await reviveMember(member, deadRole)
    } catch (e) {
      log.error('Cannot revive member:', e)
    }
  }, reviveDelayMs)
}
