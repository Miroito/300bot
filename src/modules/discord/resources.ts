import discord from 'discord.js'
import { CharacterWithOutfitWithLeader } from '@planetside/types'
import { formatWithEmojis } from './utils'

export const factionEmojis = [
  'Unknown Faction',
  '{emoji:faction_logo_vs|VS}',
  '{emoji:faction_logo_nc|NC}',
  '{emoji:faction_logo_tr|TR}',
  '{emoji:faction_logo_ns|NS}',
]

export const displayCharacter = (
  character: CharacterWithOutfitWithLeader,
  channel: discord.Channel | null,
): string => {
  return formatWithEmojis(
    channel,
    `**${character.name.first}** ${factionEmojis[Number(character.factionId)]}${
      character.factionId === '4' && character.outfitMember
        ? ` ${
            factionEmojis[
              Number(character.outfitMember.outfit.leader.factionId)
            ]
          }`
        : ''
    }`,
  )
}
