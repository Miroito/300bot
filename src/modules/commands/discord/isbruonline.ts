import { sentence } from '@app/utils/language'
import { Command } from '@commands/CommandHandler'
import { DiscordParams } from '@commands/params'
import { validateArgumentNumber } from '@commands/validators'
import { bruCharactersDatabase } from '@database/brucharacters'
import { displayCharacter } from '@discord/resources'
import { censusApi } from '@planetside/CensusApi'
import { TestDiscordParams } from '@test/utils/commands'

export default new Command<DiscordParams | TestDiscordParams>({
  keyword: 'isbruonline',
  alias: ['isbruoffline'],
  description: 'check if Bru is online',
  help: 'Usage: `{prefix}isbruonline` - checks if Bru is online',
  callback: async ({ args, alias, reply, env }) => {
    validateArgumentNumber(args.length, 0)

    const statuses =
      await censusApi.getCharacterOutfitLeaderFactionAndOnlineStatus({
        characterId: bruCharactersDatabase.keys.join(','),
      })

    if (!statuses) {
      return reply(
        'Bru has no existing characters, please verify using the brucharacters command',
      )
    }
    const channelOpt = env.message ? env.message.channel : null

    const characterList = statuses
      .filter((c) => c.onlineStatus === '10')
      .map((c) => `${displayCharacter(c, channelOpt)}`)

    const online = characterList.length > 0
    if (alias === 'isbruonline') {
      return reply(
        online
          ? `Yes, Bru is online as ${sentence(characterList)}!`
          : 'No, Bru is offline.',
      )
    } else {
      return reply(
        online
          ? `No, Bru is online as ${sentence(characterList)}.`
          : 'Yes, Bru is offline!',
      )
    }
  },
})
