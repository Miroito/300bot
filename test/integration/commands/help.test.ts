import { Command } from '@app/modules/commands/CommandHandler'
import { client } from '@app/modules/discord/client'
import { commands } from '@commands/index'
import { getMultipleCommandRunner } from '@test/utils/commands'

afterAll(() => {
  client.destroy()
})

describe('help', () => {
  const comm = commands as Command[]
  const runcommand = getMultipleCommandRunner(comm)

  it('should display the help message', async () => {
    const reply = await runcommand('+help')
    expect(reply).toEqual(
      `This is a list of all my basic commands:

**+certs** - show player's available certs
**+hello** - say hello
**+help** - show command help
**+isbruonline** - check if Bru is online
**+membercount** - count outfit members
**+playerstats** - show PS2 player stats
**+population** - display the current population
**+whoisleading** - check who is leading
**+whoisonline** - check who is online\n\n`,
    )
  })

  it('should say that there is no existing category/command', async () => {
    const reply = await runcommand('+help TH1SC0MMANDD03SNT3X1ST')
    expect(reply).toEqual('There is no catergory or command with this name.')
  })

  it('should display the full help', async () => {
    const reply = await runcommand('+help full')
    expect(reply).toEqual(
      `This is a list of all my commands:

__**Basic**__
**+certs** - show player's available certs
**+hello** - say hello
**+help** - show command help
**+isbruonline** - check if Bru is online
**+membercount** - count outfit members
**+playerstats** - show PS2 player stats
**+population** - display the current population
**+whoisleading** - check who is leading
**+whoisonline** - check who is online

__**Advanced**__
**+alts** - show player's alts
**+brudistance** - find player's interaction chain with Bru
**+directive** - show PS2 player directive
**+outfitscoreboard** - show PS2 outfit player scoreboard
**+outfitstats** - show PS2 outfit stats
**+playersearch** - find PlanetSide 2 players
**+recentstats** - show recent player stats
**+weaponstats** - show PS2 player weapon stats
**+whotokick** - show players that can be kicked from the outfit

__**Fun**__
**+coinflip** - flip a coin
**+diceroll** - roll a dice
**+hangman** - play a game of hangman
**+math** - evaluate a math expression
**+russianroulette** - kill yourself for fun purposes
**+urbandictionary** - get the modern definition of a word or expression

__**Admin**__
**+alerttracker** - configure the alert tracker
**+brucharacters** - manage Bru's characters
**+cooldown** - close a channel temporarily
**+prefix** - set command prefix
**+track** - track PlanetSide 2 characters
**+uptime** - display time since last reboot\n\n`,
    )
  })
})
