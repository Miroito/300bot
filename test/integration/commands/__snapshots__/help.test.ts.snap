// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`help should display the full help 1`] = `
"This is a list of all my commands:

__**Basic**__
**+certs** - show player's available certs
**+isbruonline** - check if Bru is online
**+membercount** - count outfit members
**+playerstats** - show PS2 player stats
**+population** - display the current population
**+whoisleading** - check who is leading
**+whoisonline** - check who is online

__**Advanced**__
**+alts** - show player's alts
**+brudistance** - find player's interaction chain with Bru
**+directive** - show PS2 player directive
**+outfitscoreboard** - show PS2 outfit player scoreboard
**+outfitstats** - show PS2 outfit stats
**+playersearch** - find PlanetSide 2 players
**+recentstats** - show recent player stats
**+uptime** - display time since last reboot
**+weaponstats** - show PS2 player weapon stats
**+whotokick** - show players that can be kicked from the outfit

__**Fun**__
**+coinflip** - flip a coin
**+diceroll** - roll a dice
**+hangman** - play a game of hangman
**+math** - evaluate a math expression
**+russianroulette** - kill yourself for fun purposes
**+urbandictionary** - get the modern definition of a word or expression

__**Admin**__
**+alerttracker** - configure the alert tracker
**+brucharacters** - manage Bru's characters
**+cooldown** - close a channel temporarily
**+prefix** - set command prefix
**+track** - track PlanetSide 2 characters

"
`;

exports[`help should display the help message 1`] = `
"This is a list of all my basic commands:

**+certs** - show player's available certs
**+isbruonline** - check if Bru is online
**+membercount** - count outfit members
**+playerstats** - show PS2 player stats
**+population** - display the current population
**+whoisleading** - check who is leading
**+whoisonline** - check who is online

"
`;
